### PR TITLE
 Add typings for sql.base, sql.compiler, and sql.ddl

### PR DESCRIPTION
**Pull request by bryanforbes**
Blocked by #17. Once it lands, I'll probably have to rebase this PR on top of it before review.

### CODE REVIEW

**Comment by bryanforbes** [comment]
This ticks the boxes for `sql/base.pyi`, `sql/compiler.pyi`, and `sql/ddl.pyi` in #3

File: sqlalchemy-stubs/sql/base.pyi [hunk 1]
```diff
@@ -1,66 +1,82 @@
...
+
 class Immutable(object):
-    def unique_params(self, *optionaldict, **kwargs): ...
-    def params(self, *optionaldict, **kwargs): ...
-    def _clone(self) -> Immutable: ...
+    def unique_params(self, *optionaldict: Any, **kwargs: Any): ...
+    def params(self, *optionaldict: Any, **kwargs: Any): ...
+    def _clone(self: _I) -> _I: ...
+
+class _DialectArgView(collections.MutableMapping[str, Any]):
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/base.pyi]: For such cases `typing.MutableMapping` (which is just an alias) is preferred.

File: sqlalchemy-stubs/sql/base.pyi [hunk 2]
```diff
@@ -1,66 +1,82 @@
...
+    def contains_column(self, col: Column) -> bool: ...
+    def as_immutable(self) -> ImmutableColumnCollection: ...
 
 class ImmutableColumnCollection(util.ImmutableProperties[Column], ColumnCollection):
-    def __init__(self, data, all_columns) -> None: ...
-    extend: Any = ...
-    remove: Any = ...
+    def __init__(self, data: Dict[str, Any], all_columns: List[Column]) -> None: ...
+    def extend(self, iter: Iterable[Column]) -> NoReturn: ...
+    def remove(self, column: Column) -> NoReturn: ...
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/base.pyi]: Are these explicitly documented or just present in code to give better error message? If the latter, then maybe just remove these two.

File: sqlalchemy-stubs/sql/ddl.pyi [hunk 3]
```diff
@@ -1,117 +1,143 @@
...
-    callable_: Any = ...
-    def execute(self, *multiparams, **params): ...
-    def execute_at(self, event_name, target): ...
-    def against(self, target): ...
+    target: Optional[SchemaItem] = ...
+    on: Optional[Union[str, Tuple[str, ...], _DDLOnCallback]] = ...
+    dialect: Optional[engine.Dialect] = ...
+    callable_: Optional[_DDLCallable] = ...
+    bind: Optional[engine.Connection] = ...
+    def execute(self, bind: Optional[engine.Connectable] = ...,  # type: ignore
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/ddl.pyi]: I would add a comment for this `# type: ignore`.

File: sqlalchemy-stubs/sql/ddl.pyi [hunk 4]
```diff
@@ -1,117 +1,143 @@
...
-    def execute_if(self, dialect: Optional[Any] = ..., callable_: Optional[Any] = ...,
-                   state: Optional[Any] = ...): ...
+    def execute_if(self: _DDLE, dialect: Optional[engine.Dialect] = ..., callable_: Optional[_DDLCallable] = ...,
+                   state: Optional[Any] = ...) -> _DDLE: ...
     def __call__(self, target, bind, **kw): ...
-    def bind(self): ...
+
+class _DDLOnCallback(Protocol):
+    def __call__(self, ddl: DDLElement, event: Optional[str], target: Optional[Union[Table, MetaData]],
+                 connection: engine.Connection, tables: Optional[List[Any]] = ...) -> bool: ...
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/ddl.pyi]: Maybe define both protocols together at the top of module? Or is this intentional?

File: sqlalchemy-stubs/sql/base.pyi (hunk 2 above)
Comment (bryanforbes) [reply sqlalchemy-stubs/sql/base.pyi]: These methods are set to `sqlalchemy.util._collections.ImmutableContainer._immutable` which throws. I imagine they can just be removed.

**Pull request merged**